File: app/src/main/java/com/nrs/nsnik/notes/util/DbUtil.java
/*
 * Copyright (C) 2017 nsnikhil
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */

package com.nrs.nsnik.notes.util;


import android.arch.lifecycle.LiveData;

import com.nrs.nsnik.notes.dagger.scopes.ApplicationScope;
import com.nrs.nsnik.notes.data.FolderEntity;
import com.nrs.nsnik.notes.data.NoteEntity;
import com.nrs.nsnik.notes.data.NotesDatabase;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Completable;
import io.reactivex.CompletableObserver;
import io.reactivex.Single;
import io.reactivex.SingleObserver;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;


@ApplicationScope
public class DbUtil {

    private final NotesDatabase mNotesDatabase;
    private final FileUtil mFileUtil;

    @Inject
    DbUtil(NotesDatabase notesDatabase, @NotNull @ApplicationScope FileUtil fileUtil) {
        this.mNotesDatabase = notesDatabase;
        this.mFileUtil = fileUtil;
    }

    public void insertNote(NoteEntity... noteEntities) {
        Single<long[]> single = Single.fromCallable(() -> mNotesDatabase.getNoteDao().insertNotes(noteEntities)).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread());
        single.subscribe(new SingleObserver<long[]>() {
            @Override
            public void onSubscribe(Disposable d) {

            }

            @Override
            public void onSuccess(long[] longs) {
                for (NoteEntity noteEntity : noteEntities) {
                    try {
                        mFileUtil.saveNote(noteEntity, noteEntity.getFileName());
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
                for (long aLong : longs) {
                    Timber.d(String.valueOf(aLong));
                }
            }

            @Override
            public void onError(Throwable e) {
                Timber.d(e.getMessage());
            }
        });
    }

    public void deleteNote(NoteEntity... noteEntities) {
        Completable completable = Completable.fromCallable(() -> {
            mNotesDatabase.getNoteDao().deleteNotes(noteEntities);
            return null;
        }).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread());
        completable.subscribe(new CompletableObserver() {
            @Override
            public void onSubscribe(Disposable d) {

            }

            @Override
            public void onComplete() {
                Timber.d("Delete successful");
            }

            @Override
            public void onError(Throwable e) {
                Timber.d(e.getMessage());
            }
        });
    }

    public void deleteNoteByFolderName(String folderName) {
        Completable completable = Completable.fromCallable(() -> {
            mNotesDatabase.getNoteDao().deleteNoteByFolderName(folderName);
            return null;
        }).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread());
        completable.subscribe(new CompletableObserver() {
            @Override
            public void onSubscribe(Disposable d) {

            }

            @Override
            public void onComplete() {
                Timber.d("Delete successful");
            }

            @Override
            public void onError(Throwable e) {
                Timber.d(e.getMessage());
            }
        });
    }

    public void updateNote(NoteEntity... noteEntities) {
        Single<Integer> single = Single.fromCallable(() -> mNotesDatabase.getNoteDao().updateNote(noteEntities)).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread());
        single.subscribe(new SingleObserver<Integer>() {
            @Override
            public void onSubscribe(Disposable d) {

            }

            @Override
            public void onSuccess(Integer integer) {
                for (NoteEntity noteEntity : noteEntities) {
                    try {
                        mFileUtil.saveNote(noteEntity, noteEntity.getFileName());
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
                Timber.d(String.valueOf(integer));
            }

            @Override
            public void onError(Throwable e) {
                Timber.d(e.getMessage());
            }
        });
    }

    public LiveData<List<NoteEntity>> getNoteList() {
        return mNotesDatabase.getNoteDao().getNotesList();
    }

    public LiveData<NoteEntity> getNoteById(int id) {
        return mNotesDatabase.getNoteDao().getNote(id);
    }

    public LiveData<List<NoteEntity>> getNoteByFolderName(String folderName) {
        return mNotesDatabase.getNoteDao().getNoteByFolderName(folderName);
    }

    public LiveData<List<NoteEntity>> getNoteByFolderNameNoPinNoLock(String folderName) {
        return mNotesDatabase.getNoteDao().getNotesByFolderNameNotPinnedNotLocked(folderName);
    }

    public LiveData<List<NoteEntity>> getNoteByFolderNamePinNoLock(String folderName) {
        return mNotesDatabase.getNoteDao().getNotesByFolderNamePinnedNotLocked(folderName);
    }

    public LiveData<List<NoteEntity>> getNoteByFolderNameNoPinLock(String folderName) {
        return mNotesDatabase.getNoteDao().getNotesByFolderNameNotPinnedLocked(folderName);
    }

    public LiveData<List<NoteEntity>> getNoteByFolderNamePinLock(String folderName) {
        return mNotesDatabase.getNoteDao().getNotesByFolderNamePinnedLocked(folderName);
    }

    public LiveData<List<NoteEntity>> searchNote(String query) {
        return mNotesDatabase.getNoteDao().getNoteByQuery(query);
    }

    public LiveData<List<NoteEntity>> getNotesByPin(int isPinned) {
        return mNotesDatabase.getNoteDao().getNoteByPin(isPinned);
    }

    public LiveData<List<NoteEntity>> getNotesByLock(int isLocked) {
        return mNotesDatabase.getNoteDao().getNoteByLock(isLocked);
    }

    public LiveData<List<NoteEntity>> getNotesByColor(String color) {
        return mNotesDatabase.getNoteDao().getNoteByColor(color);
    }

    public void insertFolder(FolderEntity... folderEntities) {
        Single<long[]> single = Single.fromCallable(() -> mNotesDatabase.getFolderDao().insertFolders(folderEntities)).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread());
        single.subscribe(new SingleObserver<long[]>() {
            @Override
            public void onSubscribe(Disposable d) {

            }

            @Override
            public void onSuccess(long[] longs) {
                for (long aLong : longs) {
                    Timber.d(String.valueOf(aLong));
                }
            }

            @Override
            public void onError(Throwable e) {
                Timber.d(e.getMessage());
            }
        });

    }

    public void deleteFolder(FolderEntity... folderEntities) {
        Completable completable = Completable.fromCallable(() -> {
            mNotesDatabase.getFolderDao().deleteFolders(folderEntities);
            return null;
        }).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread());
        completable.subscribe(new CompletableObserver() {
            @Override
            public void onSubscribe(Disposable d) {

            }

            @Override
            public void onComplete() {
                Timber.d("Delete Successful");
            }

            @Override
            public void onError(Throwable e) {
                Timber.d(e.getMessage());
            }
        });
    }

    public void deleteFolderByName(String folderName) {
        Completable completable = Completable.fromCallable(() -> {
            mNotesDatabase.getFolderDao().deleteFolderByName(folderName);
            return null;
        }).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread());
        completable.subscribe(new CompletableObserver() {
            @Override
            public void onSubscribe(Disposable d) {

            }

            @Override
            public void onComplete() {
                Timber.d("Delete Successful");
            }

            @Override
            public void onError(Throwable e) {
                Timber.d(e.getMessage());
            }
        });
    }

    public void deleteFolderByParent(String parentFolderName) {
        Completable completable = Completable.fromCallable(() -> {
            mNotesDatabase.getFolderDao().deleteFolderByParent(parentFolderName);
            return null;
        }).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread());
        completable.subscribe(new CompletableObserver() {
            @Override
            public void onSubscribe(Disposable d) {

            }

            @Override
            public void onComplete() {
                Timber.d("Delete Successful");
            }

            @Override
            public void onError(Throwable e) {
                Timber.d(e.getMessage());
            }
        });
    }

    public void updateFolder(FolderEntity... folderEntities) {
        Single<Integer> single = Single.fromCallable(() -> mNotesDatabase.getFolderDao().updateFolders(folderEntities)).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread());
        single.subscribe(new SingleObserver<Integer>() {
            @Override
            public void onSubscribe(Disposable d) {

            }

            @Override
            public void onSuccess(Integer integer) {
                Timber.d(String.valueOf(integer));
            }

            @Override
            public void onError(Throwable e) {
                Timber.d(e.getMessage());
            }
        });
    }

    public LiveData<List<FolderEntity>> getFolderList() {
        return mNotesDatabase.getFolderDao().getFoldersList();
    }

    public LiveData<FolderEntity> getFolderById(int id) {
        return mNotesDatabase.getFolderDao().getFolder(id);
    }

    public LiveData<FolderEntity> getFolderByName(String name) {
        return mNotesDatabase.getFolderDao().getFolderByName(name);
    }

    public LiveData<List<FolderEntity>> searchFolder(String query) {
        return mNotesDatabase.getFolderDao().getFolderByQuery(query);
    }

    public LiveData<List<FolderEntity>> getFolderByParent(String parentFolder) {
        return mNotesDatabase.getFolderDao().getFolderByParent(parentFolder);
    }

    public LiveData<List<FolderEntity>> getFolderByParentNoPinNoLock(String parentFolder) {
        return mNotesDatabase.getFolderDao().getFolderByParentNoPinNoLock(parentFolder);
    }

    public LiveData<List<FolderEntity>> getFolderByParentPinNoLock(String parentFolder) {
        return mNotesDatabase.getFolderDao().getFolderByParentPinNoLock(parentFolder);
    }

    public LiveData<List<FolderEntity>> getFolderByParentNoPinLock(String parentFolder) {
        return mNotesDatabase.getFolderDao().getFolderByParentNoPinLock(parentFolder);
    }

    public LiveData<List<FolderEntity>> getFolderByParentPinLock(String parentFolder) {
        return mNotesDatabase.getFolderDao().getFolderByParentPinLock(parentFolder);
    }

    public LiveData<List<FolderEntity>> getFolderByPin(int isPinned) {
        return mNotesDatabase.getFolderDao().getFolderByPin(isPinned);
    }

    public LiveData<List<FolderEntity>> getFolderByLock(int isLocked) {
        return mNotesDatabase.getFolderDao().getFolderByLock(isLocked);
    }

    public LiveData<List<FolderEntity>> getFolderByColor(String color) {
        return mNotesDatabase.getFolderDao().getFolderByColor(color);
    }
}

File: app/src/main/java/com/nrs/nsnik/notes/util/DbUtil.kt
/*
 *     Notes  Copyright (C) 2018  Nikhil Soni
 *     This program comes with ABSOLUTELY NO WARRANTY; for details type `show w'.
 *     This is free software, and you are welcome to redistribute it
 *     under certain conditions; type `show c' for details.
 *
 * The hypothetical commands `show w' and `show c' should show the appropriate
 * parts of the General Public License.  Of course, your program's commands
 * might be different; for a GUI interface, you would use an "about box".
 *
 *   You should also get your employer (if you work as a programmer) or school,
 * if any, to sign a "copyright disclaimer" for the program, if necessary.
 * For more information on this, and how to apply and follow the GNU GPL, see
 * <http://www.gnu.org/licenses/>.
 *
 *   The GNU General Public License does not permit incorporating your program
 * into proprietary programs.  If your program is a subroutine library, you
 * may consider it more useful to permit linking proprietary applications with
 * the library.  If this is what you want to do, use the GNU Lesser General
 * Public License instead of this License.  But first, please read
 * <http://www.gnu.org/philosophy/why-not-lgpl.html>.
 */

package com.nrs.nsnik.notes.util


import androidx.lifecycle.LiveData
import com.nrs.nsnik.notes.dagger.scopes.ApplicationScope
import com.nrs.nsnik.notes.data.FolderEntity
import com.nrs.nsnik.notes.data.NoteEntity
import com.nrs.nsnik.notes.data.NotesDatabase
import io.reactivex.Completable
import io.reactivex.CompletableObserver
import io.reactivex.Single
import io.reactivex.SingleObserver
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import timber.log.Timber
import java.io.IOException
import javax.inject.Inject


@ApplicationScope
class DbUtil @Inject
internal constructor(private val mNotesDatabase: NotesDatabase, @param:ApplicationScope private val mFileUtil: FileUtil) {

    val noteList: LiveData<List<NoteEntity>>
        get() = mNotesDatabase.noteDao.notesList

    val folderList: LiveData<List<FolderEntity>>
        get() = mNotesDatabase.folderDao.foldersList

    fun insertNote(vararg noteEntities: NoteEntity) {
        val single = Single.fromCallable { mNotesDatabase.noteDao.insertNotes(*noteEntities) }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
        single.subscribe(object : SingleObserver<LongArray> {
            override fun onSubscribe(d: Disposable) {

            }

            override fun onSuccess(longs: LongArray) {
                noteEntities.forEach {
                    try {
                        mFileUtil.saveNote(it, it.fileName!!)
                    } catch (e: IOException) {
                        e.printStackTrace()
                    }

                }
                longs.forEach {
                    Timber.d(it.toString())
                }
            }

            override fun onError(e: Throwable) {
                Timber.d(e.message)
            }
        })
    }

    fun deleteNote(vararg noteEntities: NoteEntity) {
        val completable = Completable.fromCallable {
            mNotesDatabase.noteDao.deleteNotes(*noteEntities)
            null
        }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
        completable.subscribe(object : CompletableObserver {
            override fun onSubscribe(d: Disposable) {

            }

            override fun onComplete() {
                Timber.d("Delete successful")
            }

            override fun onError(e: Throwable) {
                Timber.d(e.message)
            }
        })
    }

    fun deleteNoteByFolderName(folderName: String) {
        val completable = Completable.fromCallable {
            mNotesDatabase.noteDao.deleteNoteByFolderName(folderName)
            null
        }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
        completable.subscribe(object : CompletableObserver {
            override fun onSubscribe(d: Disposable) {

            }

            override fun onComplete() {
                Timber.d("Delete successful")
            }

            override fun onError(e: Throwable) {
                Timber.d(e.message)
            }
        })
    }

    fun updateNote(vararg noteEntities: NoteEntity) {
        val single = Single.fromCallable { mNotesDatabase.noteDao.updateNote(*noteEntities) }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
        single.subscribe(object : SingleObserver<Int> {
            override fun onSubscribe(d: Disposable) {

            }

            override fun onSuccess(t: Int) {
                noteEntities.forEach {
                    try {
                        mFileUtil.saveNote(it, it.fileName!!)
                    } catch (e: IOException) {
                        e.printStackTrace()
                    }
                }
                Timber.d(t.toString())
            }

            override fun onError(e: Throwable) {
                Timber.d(e.message)
            }
        })
    }

    fun getNoteById(id: Int): LiveData<NoteEntity> {
        return mNotesDatabase.noteDao.getNote(id)
    }

    fun getNoteByFolderName(folderName: String): LiveData<List<NoteEntity>> {
        return mNotesDatabase.noteDao.getNoteByFolderName(folderName)
    }

    fun getNoteByFolderNameNoPinNoLock(folderName: String): LiveData<List<NoteEntity>> {
        return mNotesDatabase.noteDao.getNotesByFolderNameNotPinnedNotLocked(folderName)
    }

    fun getNoteByFolderNamePinNoLock(folderName: String): LiveData<List<NoteEntity>> {
        return mNotesDatabase.noteDao.getNotesByFolderNamePinnedNotLocked(folderName)
    }

    fun getNoteByFolderNameNoPinLock(folderName: String): LiveData<List<NoteEntity>> {
        return mNotesDatabase.noteDao.getNotesByFolderNameNotPinnedLocked(folderName)
    }

    fun getNoteByFolderNamePinLock(folderName: String): LiveData<List<NoteEntity>> {
        return mNotesDatabase.noteDao.getNotesByFolderNamePinnedLocked(folderName)
    }

    fun getNoteByFolderNameOrdered(folderName: String): LiveData<List<NoteEntity>> {
        return mNotesDatabase.noteDao.getNoteByFolderNameOrdered(folderName)
    }

    fun getNoteByFolderNameOrderedLock(folderName: String): LiveData<List<NoteEntity>> {
        return mNotesDatabase.noteDao.getNoteByFolderNameOrderedLock(folderName)
    }

    fun searchNote(query: String): LiveData<List<NoteEntity>> {
        return mNotesDatabase.noteDao.getNoteByQuery(query)
    }

    fun getNotesByPin(isPinned: Int): LiveData<List<NoteEntity>> {
        return mNotesDatabase.noteDao.getNoteByPin(isPinned)
    }

    fun getNotesByLock(isLocked: Int): LiveData<List<NoteEntity>> {
        return mNotesDatabase.noteDao.getNoteByLock(isLocked)
    }

    fun getNotesByColor(color: String): LiveData<List<NoteEntity>> {
        return mNotesDatabase.noteDao.getNoteByColor(color)
    }

    fun changeNotePinStatus(id: Int, pin: Int) {
        val single = Single.fromCallable {
            mNotesDatabase.noteDao.changeNotePinStatus(id, pin)
        }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
        single.subscribe(object : SingleObserver<Unit> {
            override fun onSubscribe(d: Disposable) {

            }

            override fun onSuccess(t: Unit) {
                Timber.d("Updated")
            }

            override fun onError(e: Throwable) {
                Timber.d(e.message)
            }
        })
    }

    fun changeNoteLockStatus(id: Int, lock: Int) {
        val single = Single.fromCallable {
            mNotesDatabase.noteDao.changeNoteLockStatus(id, lock)
        }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
        single.subscribe(object : SingleObserver<Unit> {
            override fun onSubscribe(d: Disposable) {

            }

            override fun onSuccess(t: Unit) {
                Timber.d("Updated")
            }

            override fun onError(e: Throwable) {
                Timber.d(e.message)
            }
        })
    }


    fun changeNoteFolder(id: Int, folderName: String) {
        val single = Single.fromCallable {
            mNotesDatabase.noteDao.changeNoteFolder(id, folderName)
        }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
        single.subscribe(object : SingleObserver<Unit> {
            override fun onSubscribe(d: Disposable) {

            }

            override fun onSuccess(t: Unit) {
                Timber.d("Updated")
            }

            override fun onError(e: Throwable) {
                Timber.d(e.message)
            }
        })
    }

    /**
     * FOLDER
     */

    fun insertFolder(vararg folderEntities: FolderEntity) {
        val single = Single.fromCallable {
            mNotesDatabase.folderDao.insertFolders(*folderEntities)
        }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
        single.subscribe(object : SingleObserver<LongArray> {
            override fun onSubscribe(d: Disposable) {

            }

            override fun onSuccess(longs: LongArray) {
                longs.forEach {
                    Timber.d(it.toString())
                }
            }

            override fun onError(e: Throwable) {
                Timber.d(e.message)
            }
        })

    }

    fun deleteFolder(vararg folderEntities: FolderEntity) {
        val completable = Completable.fromCallable {
            mNotesDatabase.folderDao.deleteFolders(*folderEntities)
            null
        }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
        completable.subscribe(object : CompletableObserver {
            override fun onSubscribe(d: Disposable) {

            }

            override fun onComplete() {
                Timber.d("Delete Successful")
            }

            override fun onError(e: Throwable) {
                Timber.d(e.message)
            }
        })
    }

    fun deleteFolderByName(folderName: String) {
        val completable = Completable.fromCallable {
            mNotesDatabase.folderDao.deleteFolderByName(folderName)
            null
        }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
        completable.subscribe(object : CompletableObserver {
            override fun onSubscribe(d: Disposable) {

            }

            override fun onComplete() {
                Timber.d("Delete Successful")
            }

            override fun onError(e: Throwable) {
                Timber.d(e.message)
            }
        })
    }

    fun deleteFolderByParent(parentFolderName: String) {
        val completable = Completable.fromCallable {
            mNotesDatabase.folderDao.deleteFolderByParent(parentFolderName)
            null
        }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
        completable.subscribe(object : CompletableObserver {
            override fun onSubscribe(d: Disposable) {

            }

            override fun onComplete() {
                Timber.d("Delete Successful")
            }

            override fun onError(e: Throwable) {
                Timber.d(e.message)
            }
        })
    }

    fun updateFolder(vararg folderEntities: FolderEntity) {
        val single = Single.fromCallable { mNotesDatabase.folderDao.updateFolders(*folderEntities) }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
        single.subscribe(object : SingleObserver<Int> {
            override fun onSubscribe(d: Disposable) {

            }

            override fun onSuccess(t: Int) {
                Timber.d(t.toString())
            }

            override fun onError(e: Throwable) {
                Timber.d(e.message)
            }
        })
    }

    fun getFolders(): LiveData<List<FolderEntity>> {
        return mNotesDatabase.folderDao.getFolders()
    }

    fun getFolderById(id: Int): LiveData<FolderEntity> {
        return mNotesDatabase.folderDao.getFolder(id)
    }

    fun getFolderByName(name: String): LiveData<FolderEntity> {
        return mNotesDatabase.folderDao.getFolderByName(name)
    }

    fun searchFolder(query: String): LiveData<List<FolderEntity>> {
        return mNotesDatabase.folderDao.getFolderByQuery(query)
    }

    fun getFolderByParent(parentFolder: String): LiveData<List<FolderEntity>> {
        return mNotesDatabase.folderDao.getFolderByParent(parentFolder)
    }

    fun getFolderByParentNoPinNoLock(parentFolder: String): LiveData<List<FolderEntity>> {
        return mNotesDatabase.folderDao.getFolderByParentNoPinNoLock(parentFolder)
    }

    fun getFolderByParentPinNoLock(parentFolder: String): LiveData<List<FolderEntity>> {
        return mNotesDatabase.folderDao.getFolderByParentPinNoLock(parentFolder)
    }

    fun getFolderByParentNoPinLock(parentFolder: String): LiveData<List<FolderEntity>> {
        return mNotesDatabase.folderDao.getFolderByParentNoPinLock(parentFolder)
    }

    fun getFolderByParentPinLock(parentFolder: String): LiveData<List<FolderEntity>> {
        return mNotesDatabase.folderDao.getFolderByParentPinLock(parentFolder)
    }

    fun getFolderByParentOrdered(parentFolder: String): LiveData<List<FolderEntity>> {
        return mNotesDatabase.folderDao.getFolderByParentOrdered(parentFolder)
    }

    fun getFolderByParentOrderedLock(parentFolder: String): LiveData<List<FolderEntity>> {
        return mNotesDatabase.folderDao.getFolderByParentOrderedLock(parentFolder)
    }

    fun getFolderByPin(isPinned: Int): LiveData<List<FolderEntity>> {
        return mNotesDatabase.folderDao.getFolderByPin(isPinned)
    }

    fun getFolderByLock(isLocked: Int): LiveData<List<FolderEntity>> {
        return mNotesDatabase.folderDao.getFolderByLock(isLocked)
    }

    fun getFolderByColor(color: String): LiveData<List<FolderEntity>> {
        return mNotesDatabase.folderDao.getFolderByColor(color)
    }

    fun changeFolderPinStatus(id: Int, pin: Int) {
        val single = Single.fromCallable {
            mNotesDatabase.folderDao.changeFolderPinStatus(id, pin)
        }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
        single.subscribe(object : SingleObserver<Unit> {
            override fun onSubscribe(d: Disposable) {

            }

            override fun onSuccess(t: Unit) {
                Timber.d("Updated")
            }

            override fun onError(e: Throwable) {
                Timber.d(e.message)
            }
        })
    }

    fun changeFolderLockStatus(id: Int, lock: Int) {
        val single = Single.fromCallable {
            mNotesDatabase.folderDao.changeFolderLockStatus(id, lock)
        }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
        single.subscribe(object : SingleObserver<Unit> {
            override fun onSubscribe(d: Disposable) {

            }

            override fun onSuccess(t: Unit) {
                Timber.d("Updated")
            }

            override fun onError(e: Throwable) {
                Timber.d(e.message)
            }
        })
    }

    fun changeFolderParent(id: Int, parentFolderName: String) {
        val single = Single.fromCallable {
            mNotesDatabase.folderDao.changeFolderParent(id, parentFolderName)
        }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
        single.subscribe(object : SingleObserver<Unit> {
            override fun onSubscribe(d: Disposable) {

            }

            override fun onSuccess(t: Unit) {
                Timber.d("Updated")
            }

            override fun onError(e: Throwable) {
                Timber.d(e.message)
            }
        })
    }


}
